### PR TITLE
Prevent setting attributes on `KerasTensor`

### PR DESCRIPTION
**Pull request by james77777778**
This PR should improve consistency when using `KerasTensor`

### CODE REVIEW

**Review by fchollet — COMMENTED** [review]
Thanks for the PR. In what cases would we set these attributes? Shoudn't these properties be immutable, like they are for a concrete tensor?

**Comment by james77777778** [comment]
> Thanks for the PR. In what cases would we set these attributes? Shouldn't these properties be immutable, like they are for a concrete tensor?

Sometimes, it is convenient to change them inplace in `compute_output_spec`:
https://github.com/keras-team/keras/blob/d318fc7d2e788c5eced2eae7d131958f488def11/keras/src/ops/core.py#L105-L119
https://github.com/keras-team/keras/blob/d318fc7d2e788c5eced2eae7d131958f488def11/keras/src/dtype_policies/dtype_policy.py#L162-L165

Should this usage be considered inappropriate?

**Comment by fchollet** [comment]
> Or should these be considered inappropriate?

I would say it's better to create a new `KerasTensor` instance for these (the existing instance might be in use elsewhere). I'd expect these properties to be immutable.

**Comment by james77777778** [comment]
Got it. Let me try to fix them.

File: keras/src/dtype_policies/dtype_policy.py [hunk 1]
```diff
@@ -161,7 +162,7 @@ def convert_input(self, x, autocast, dtype):
             return x
         elif backend.is_keras_tensor(x):
             if self._should_cast(x, autocast, dtype):
-                x.dtype = dtype
+                x = ops.Cast(dtype=dtype)(x)
```
Comment (fchollet) [inline keras/src/dtype_policies/dtype_policy.py:165]: Note that just `ops.cast` would work.

File: keras/src/ops/core.py [hunk 2]
```diff
@@ -114,9 +113,9 @@ def compute_output_spec(self, f, init, xs, length):
             )
             x = xs[0]
 
-        carry_spec, y_spec = backend.compute_output_spec(f, init, x)
-        y_spec.shape = (n,) + y_spec.shape
-        return carry_spec, y_spec
+        carry, y = backend.compute_output_spec(f, init, x)
+        y = KerasTensor((n,) + y.shape, y.dtype, y.sparse)
```
Comment (fchollet) [inline keras/src/ops/core.py:117]: Prefer using keyword args for better maintainability -- `dtype=y.dtype, sparse=y.sparse`

**Review by fchollet — COMMENTED** [review]
Thanks for the update!

File: keras/src/dtype_policies/dtype_policy.py (hunk 1 above)
Comment (james77777778) [reply keras/src/dtype_policies/dtype_policy.py]: Fixed.

File: keras/src/ops/core.py (hunk 2 above)
Comment (james77777778) [reply keras/src/ops/core.py]: Fixed.

**Comment by james77777778** [comment]
I've added a more clearer error msg for the setters of KerasTensor.
This should guide anyone who tries to modify these attributes to create a new instance.

**Review by fchollet — APPROVED** [review]
LGTM, thank you!

**Pull request merged**